File: javascript/src/api/operational_webhook_endpoint.ts
// this file is @generated
import {
  ListResponseOperationalWebhookEndpointOut,
  ListResponseOperationalWebhookEndpointOutUtil,
} from "../models/list_response_operational_webhook_endpoint_out";
import {
  OperationalWebhookEndpointHeadersIn,
  OperationalWebhookEndpointHeadersInUtil,
} from "../models/operational_webhook_endpoint_headers_in";
import {
  OperationalWebhookEndpointHeadersOut,
  OperationalWebhookEndpointHeadersOutUtil,
} from "../models/operational_webhook_endpoint_headers_out";
import {
  OperationalWebhookEndpointIn,
  OperationalWebhookEndpointInUtil,
} from "../models/operational_webhook_endpoint_in";
import {
  OperationalWebhookEndpointOut,
  OperationalWebhookEndpointOutUtil,
} from "../models/operational_webhook_endpoint_out";
import {
  OperationalWebhookEndpointSecretIn,
  OperationalWebhookEndpointSecretInUtil,
} from "../models/operational_webhook_endpoint_secret_in";
import {
  OperationalWebhookEndpointSecretOut,
  OperationalWebhookEndpointSecretOutUtil,
} from "../models/operational_webhook_endpoint_secret_out";
import {
  OperationalWebhookEndpointUpdate,
  OperationalWebhookEndpointUpdateUtil,
} from "../models/operational_webhook_endpoint_update";
import { Ordering, OrderingUtil } from "../models/ordering";
import { HttpMethod, SvixRequest, SvixRequestContext } from "../request";

export interface OperationalWebhookEndpointListOptions {
  /** Limit the number of returned items */
  limit?: number;
  /** The iterator returned from a prior invocation */
  iterator?: string | null;
  /** The sorting order of the returned items */
  order?: Ordering;
}

export interface OperationalWebhookEndpointCreateOptions {
  idempotencyKey?: string;
}

export interface OperationalWebhookEndpointRotateSecretOptions {
  idempotencyKey?: string;
}

export class OperationalWebhookEndpoint {
  public constructor(private readonly requestCtx: SvixRequestContext) {}

  /** List operational webhook endpoints. */
  public list(
    options?: OperationalWebhookEndpointListOptions
  ): Promise<ListResponseOperationalWebhookEndpointOut> {
    const request = new SvixRequest(
      HttpMethod.GET,
      "/api/v1/operational-webhook/endpoint"
    );

    request.setQueryParam("limit", options?.limit);
    request.setQueryParam("iterator", options?.iterator);
    request.setQueryParam("order", options?.order);

    return request.send(
      this.requestCtx,
      ListResponseOperationalWebhookEndpointOutUtil._fromJsonObject
    );
  }

  /** Create an operational webhook endpoint. */
  public create(
    operationalWebhookEndpointIn: OperationalWebhookEndpointIn,
    options?: OperationalWebhookEndpointCreateOptions
  ): Promise<OperationalWebhookEndpointOut> {
    const request = new SvixRequest(
      HttpMethod.POST,
      "/api/v1/operational-webhook/endpoint"
    );

    request.setHeaderParam("idempotency-key", options?.idempotencyKey);
    request.setBody(
      OperationalWebhookEndpointInUtil._toJsonObject(operationalWebhookEndpointIn)
    );

    return request.send(
      this.requestCtx,
      OperationalWebhookEndpointOutUtil._fromJsonObject
    );
  }

  /** Get an operational webhook endpoint. */
  public get(endpointId: string): Promise<OperationalWebhookEndpointOut> {
    const request = new SvixRequest(
      HttpMethod.GET,
      "/api/v1/operational-webhook/endpoint/{endpoint_id}"
    );

    request.setPathParam("endpoint_id", endpointId);

    return request.send(
      this.requestCtx,
      OperationalWebhookEndpointOutUtil._fromJsonObject
    );
  }

  /** Update an operational webhook endpoint. */
  public update(
    endpointId: string,
    operationalWebhookEndpointUpdate: OperationalWebhookEndpointUpdate
  ): Promise<OperationalWebhookEndpointOut> {
    const request = new SvixRequest(
      HttpMethod.PUT,
      "/api/v1/operational-webhook/endpoint/{endpoint_id}"
    );

    request.setPathParam("endpoint_id", endpointId);
    request.setBody(
      OperationalWebhookEndpointUpdateUtil._toJsonObject(operationalWebhookEndpointUpdate)
    );

    return request.send(
      this.requestCtx,
      OperationalWebhookEndpointOutUtil._fromJsonObject
    );
  }

  /** Delete an operational webhook endpoint. */
  public delete(endpointId: string): Promise<void> {
    const request = new SvixRequest(
      HttpMethod.DELETE,
      "/api/v1/operational-webhook/endpoint/{endpoint_id}"
    );

    request.setPathParam("endpoint_id", endpointId);

    return request.sendNoResponseBody(this.requestCtx);
  }

  /** Get the additional headers to be sent with the operational webhook. */
  public getHeaders(endpointId: string): Promise<OperationalWebhookEndpointHeadersOut> {
    const request = new SvixRequest(
      HttpMethod.GET,
      "/api/v1/operational-webhook/endpoint/{endpoint_id}/headers"
    );

    request.setPathParam("endpoint_id", endpointId);

    return request.send(
      this.requestCtx,
      OperationalWebhookEndpointHeadersOutUtil._fromJsonObject
    );
  }

  /** Set the additional headers to be sent with the operational webhook. */
  public updateHeaders(
    endpointId: string,
    operationalWebhookEndpointHeadersIn: OperationalWebhookEndpointHeadersIn
  ): Promise<void> {
    const request = new SvixRequest(
      HttpMethod.PUT,
      "/api/v1/operational-webhook/endpoint/{endpoint_id}/headers"
    );

    request.setPathParam("endpoint_id", endpointId);
    request.setBody(
      OperationalWebhookEndpointHeadersInUtil._toJsonObject(
        operationalWebhookEndpointHeadersIn
      )
    );

    return request.sendNoResponseBody(this.requestCtx);
  }

  /**
   * Get an operational webhook endpoint's signing secret.
   *
   * This is used to verify the authenticity of the webhook.
   * For more information please refer to [the consuming webhooks docs](https://docs.svix.com/consuming-webhooks/).
   */
  public getSecret(endpointId: string): Promise<OperationalWebhookEndpointSecretOut> {
    const request = new SvixRequest(
      HttpMethod.GET,
      "/api/v1/operational-webhook/endpoint/{endpoint_id}/secret"
    );

    request.setPathParam("endpoint_id", endpointId);

    return request.send(
      this.requestCtx,
      OperationalWebhookEndpointSecretOutUtil._fromJsonObject
    );
  }

  /**
   * Rotates an operational webhook endpoint's signing secret.
   *
   * The previous secret will remain valid for the next 24 hours.
   */
  public rotateSecret(
    endpointId: string,
    operationalWebhookEndpointSecretIn: OperationalWebhookEndpointSecretIn,
    options?: OperationalWebhookEndpointRotateSecretOptions
  ): Promise<void> {
    const request = new SvixRequest(
      HttpMethod.POST,
      "/api/v1/operational-webhook/endpoint/{endpoint_id}/secret/rotate"
    );

    request.setPathParam("endpoint_id", endpointId);
    request.setHeaderParam("idempotency-key", options?.idempotencyKey);
    request.setBody(
      OperationalWebhookEndpointSecretInUtil._toJsonObject(
        operationalWebhookEndpointSecretIn
      )
    );

    return request.sendNoResponseBody(this.requestCtx);
  }
}

File: kotlin/lib/src/main/kotlin/OperationalWebhookEndpoint.kt
// this file is @generated
package com.svix.kotlin

import com.svix.kotlin.models.ListResponseOperationalWebhookEndpointOut
import com.svix.kotlin.models.OperationalWebhookEndpointHeadersIn
import com.svix.kotlin.models.OperationalWebhookEndpointHeadersOut
import com.svix.kotlin.models.OperationalWebhookEndpointIn
import com.svix.kotlin.models.OperationalWebhookEndpointOut
import com.svix.kotlin.models.OperationalWebhookEndpointSecretIn
import com.svix.kotlin.models.OperationalWebhookEndpointSecretOut
import com.svix.kotlin.models.OperationalWebhookEndpointUpdate
import com.svix.kotlin.models.Ordering
import okhttp3.Headers
import okhttp3.HttpUrl

data class OperationalWebhookEndpointListOptions(
    val limit: ULong? = null,
    val iterator: String? = null,
    val order: Ordering? = null,
)

data class OperationalWebhookEndpointCreateOptions(val idempotencyKey: String? = null)

data class OperationalWebhookEndpointRotateSecretOptions(val idempotencyKey: String? = null)

class OperationalWebhookEndpoint(baseUrl: HttpUrl, defaultHeaders: Map<String, String>) :
    SvixHttpClient(baseUrl, defaultHeaders) {

    /** List operational webhook endpoints. */
    suspend fun list(
        options: OperationalWebhookEndpointListOptions = OperationalWebhookEndpointListOptions()
    ): ListResponseOperationalWebhookEndpointOut {
        var url = this.newUrlBuilder().encodedPath("/api/v1/operational-webhook/endpoint")
        options.limit?.let { url = url.addQueryParameter("limit", it.toString()) }
        options.iterator?.let { url = url.addQueryParameter("iterator", it) }
        options.order?.let { url = url.addQueryParameter("order", it.toString()) }
        return this.executeRequest<Any, ListResponseOperationalWebhookEndpointOut>(
            "GET",
            url.build(),
        )
    }

    /** Create an operational webhook endpoint. */
    suspend fun create(
        operationalWebhookEndpointIn: OperationalWebhookEndpointIn,
        options: OperationalWebhookEndpointCreateOptions = OperationalWebhookEndpointCreateOptions(),
    ): OperationalWebhookEndpointOut {
        val url = this.newUrlBuilder().encodedPath("/api/v1/operational-webhook/endpoint")
        var headers = Headers.Builder()
        options.idempotencyKey?.let { headers = headers.add("idempotency-key", it) }

        return this.executeRequest<OperationalWebhookEndpointIn, OperationalWebhookEndpointOut>(
            "POST",
            url.build(),
            headers = headers.build(),
            reqBody = operationalWebhookEndpointIn,
        )
    }

    /** Get an operational webhook endpoint. */
    suspend fun get(endpointId: String): OperationalWebhookEndpointOut {
        val url =
            this.newUrlBuilder().encodedPath("/api/v1/operational-webhook/endpoint/$endpointId")
        return this.executeRequest<Any, OperationalWebhookEndpointOut>("GET", url.build())
    }

    /** Update an operational webhook endpoint. */
    suspend fun update(
        endpointId: String,
        operationalWebhookEndpointUpdate: OperationalWebhookEndpointUpdate,
    ): OperationalWebhookEndpointOut {
        val url =
            this.newUrlBuilder().encodedPath("/api/v1/operational-webhook/endpoint/$endpointId")

        return this.executeRequest<OperationalWebhookEndpointUpdate, OperationalWebhookEndpointOut>(
            "PUT",
            url.build(),
            reqBody = operationalWebhookEndpointUpdate,
        )
    }

    /** Delete an operational webhook endpoint. */
    suspend fun delete(endpointId: String) {
        val url =
            this.newUrlBuilder().encodedPath("/api/v1/operational-webhook/endpoint/$endpointId")
        this.executeRequest<Any, Boolean>("DELETE", url.build())
    }

    /** Get the additional headers to be sent with the operational webhook. */
    suspend fun getHeaders(endpointId: String): OperationalWebhookEndpointHeadersOut {
        val url =
            this.newUrlBuilder()
                .encodedPath("/api/v1/operational-webhook/endpoint/$endpointId/headers")
        return this.executeRequest<Any, OperationalWebhookEndpointHeadersOut>("GET", url.build())
    }

    /** Set the additional headers to be sent with the operational webhook. */
    suspend fun updateHeaders(
        endpointId: String,
        operationalWebhookEndpointHeadersIn: OperationalWebhookEndpointHeadersIn,
    ) {
        val url =
            this.newUrlBuilder()
                .encodedPath("/api/v1/operational-webhook/endpoint/$endpointId/headers")

        this.executeRequest<OperationalWebhookEndpointHeadersIn, Boolean>(
            "PUT",
            url.build(),
            reqBody = operationalWebhookEndpointHeadersIn,
        )
    }

    /**
     * Get an operational webhook endpoint's signing secret.
     *
     * This is used to verify the authenticity of the webhook. For more information please refer to
     * [the consuming webhooks docs](https://docs.svix.com/consuming-webhooks/).
     */
    suspend fun getSecret(endpointId: String): OperationalWebhookEndpointSecretOut {
        val url =
            this.newUrlBuilder()
                .encodedPath("/api/v1/operational-webhook/endpoint/$endpointId/secret")
        return this.executeRequest<Any, OperationalWebhookEndpointSecretOut>("GET", url.build())
    }

    /**
     * Rotates an operational webhook endpoint's signing secret.
     *
     * The previous secret will remain valid for the next 24 hours.
     */
    suspend fun rotateSecret(
        endpointId: String,
        operationalWebhookEndpointSecretIn: OperationalWebhookEndpointSecretIn,
        options: OperationalWebhookEndpointRotateSecretOptions =
            OperationalWebhookEndpointRotateSecretOptions(),
    ) {
        val url =
            this.newUrlBuilder()
                .encodedPath("/api/v1/operational-webhook/endpoint/$endpointId/secret/rotate")
        var headers = Headers.Builder()
        options.idempotencyKey?.let { headers = headers.add("idempotency-key", it) }

        this.executeRequest<OperationalWebhookEndpointSecretIn, Boolean>(
            "POST",
            url.build(),
            headers = headers.build(),
            reqBody = operationalWebhookEndpointSecretIn,
        )
    }
}
